lib/bots/keyframebot.ts

```diff
@@ -0,0 +1,683 @@
+                        number: prNumber,
+                    },
+                    method: this.githubApi.issues.createComment,
+                });
+
+            }
+            return commentPromise.then(() => {
+                return this.dispatchToEmitter(this.githubEmitterName, {
+                    data: {
+                        context: 'KeyframeBot',
+                        description: lintMessage,
+                        owner,
+                        repo,
+                        sha: head.sha,
+                        state: (lintResults.valid) ? 'success' : 'failure'
+                    },
+                    method: this.githubApi.repos.createStatus,
+                });
+            });
+        }).finally(tempCleanup);
+    }
+
+    /**
+     * Deployment route method, hanging off an Express server.
+     * When a valid keyframe version and environment is passed, this will verify the product keyframe,
+     * and then create a new branch in the environment. The keyframe will be committed to this branch
+     * and a new PR will be based off it. This will automatically be linted by the process.
+     * @param req   The HTTP request.
+     * @param res   The HTTP response.
+     */
+    private deployKeyframe = (req: express.Request, res: express.Response): void => {
+        const payload: KeyframeDetails = req.body;
+        const environment = payload.environment;
+        const version = payload.version;
+        const headerToken = req.get('Authorization');
+        const productSplitRepo = this.productRepo.split('/');
+        const productOwner = productSplitRepo[0];
+        const productRepo = productSplitRepo[1];
+        const resin = resinSdk();
+
+        let decodedToken: any;
+        let owner = '';
+        let repo = '';
+        let deployDetails: DeploymentDetails;
+
+        // Read the headers, validate the bearer token with the SDK.
+        const tokenMatch = headerToken.match(/^token (.*)$/i);
+        if (!tokenMatch) {
+            res.sendStatus(400);
+            return;
+        }
+
+        const token = tokenMatch[1];
+        resin.auth.loginWithToken(token).then(() => {
+            try {
+                decodedToken = jwtDecode(token);
+            } catch(_err) {
+                throw new Error('Cannot decode token into JWT object');
+            }
+
+            if (!_.includes(decodedToken.permissions, 'admin.home')) {
+                // Ensure it's a 404 so anyone without rights doesn't know it exists.
+                throw new HTTPError(401, 'Invalid access rights');
+            }
+
+            // Get the right environment.
+            const envRepo = this.environments[environment];
+            if (!envRepo) {
+                throw new HTTPError(404, 'Passed environment does not exist');
+            }
+
+            // Prep for the environment PR.
+            const splitRepo = envRepo.split('/');
+            owner = splitRepo[0];
+            repo = splitRepo[1];
```

Comment:
Fair enough. I guess it depends on the ratio of the number of methods that take `owner` and `repo` as separate parameters (which I guess is all the 'raw' GH API calls) to the number of methods that could be modified to take a single OwnerRepoObject (potentially every other ProcBot method). I guess it also depends on the level of abstraction that you're intending to target with the ProcBots framework.
\*shrug\*